dspy/clients/databricks.py

```diff
@@ -0,0 +1,326 @@
+                    )
+                elif data_format == DataFormat.completion:
+                    client.completions.create(prompt="hi", model=model_name, max_tokens=1)
+                return
+            except Exception:
+                time.sleep(60)
+
+    @staticmethod
+    def finetune(
+        job: TrainingJobDatabricks,
+        model: str,
+        train_data: List[Dict[str, Any]],
+        train_kwargs: Optional[Dict[str, Any]] = None,
+        data_format: Optional[Union[DataFormat, str]] = None,
+    ) -> str:
+        if isinstance(data_format, str):
+            if data_format == "chat":
+                data_format = DataFormat.chat
+            elif data_format == "completion":
+                data_format = DataFormat.completion
+            else:
+                raise ValueError(
+                    f"String `data_format` must be one of 'chat' or 'completion', but received: {data_format}."
+                )
+
+        if "train_data_path" not in train_kwargs:
+            raise ValueError("The `train_data_path` must be provided to finetune on Databricks.")
+        # Add the file name to the directory path.
+        train_kwargs["train_data_path"] = DatabricksProvider.upload_data(
+            train_data, train_kwargs["train_data_path"], data_format
+        )
+
+        try:
+            from databricks.model_training import foundation_model as fm
+        except ImportError:
+            raise ImportError(
+                "To use Databricks finetuning, please install the databricks_genai package via "
+                "`pip install databricks_genai`."
+            )
+
+        if "register_to" not in train_kwargs:
+            raise ValueError("The `register_to` must be provided to finetune on Databricks.")
+
+        # Allow users to override the host and token. This is useful on Databricks hosted runtime.
+        databricks_host = train_kwargs.pop("databricks_host", None)
+        databricks_token = train_kwargs.pop("databricks_token", None)
+
+        finetuning_run = fm.create(
+            model=model,
+            **train_kwargs,
+        )
+
+        job.run = finetuning_run
+
+        # Wait for the finetuning run to be ready.
+        while True:
+            job.run = fm.get(job.run)
+            if job.run.status.display_name == "Succeeded":
+                break
+            elif job.run.status.display_name == "Failed":
+                raise ValueError(
+                    f"Finetuning run failed with status: {job.run.status.display_name}. Please check the Databricks "
+                    f"workspace for more details. Finetuning job's metadata: {job.run}。"
+                )
+            else:
+                time.sleep(60)
+
+        job.launch_started = True
+        model_to_deploy = train_kwargs.get("register_to")
+        job.endpoint_name = model_to_deploy.replace(".", "_")
+        DatabricksProvider.deploy_finetuned_model(model_to_deploy, data_format, databricks_host, databricks_token)
```

Comment:
Should this be here or inside launch?

Reply:
Ideally this should, but the current implementation of LM and optimizer requires `finetune()` to return a ready-to-use endpoint. This is actually one thing I would like to improve, but I decided to stick to the current design to speed things up. 

I will work on the general improvements during my side time afterwards.